pkg/virt-controller/watch/ovm.go

```diff
@@ -0,0 +1,695 @@
+	}
+}
+
+func (c *OVMController) Execute() bool {
+	key, quit := c.Queue.Get()
+	if quit {
+		return false
+	}
+	defer c.Queue.Done(key)
+	if err := c.execute(key.(string)); err != nil {
+		log.Log.Reason(err).Infof("re-enqueuing OfflineVirtualMachine %v", key)
+		c.Queue.AddRateLimited(key)
+	} else {
+		log.Log.V(4).Infof("processed OfflineVirtualMachine %v", key)
+		c.Queue.Forget(key)
+	}
+	return true
+}
+
+func (c *OVMController) execute(key string) error {
+
+	obj, exists, err := c.vmOVMInformer.GetStore().GetByKey(key)
+	if err != nil {
+		return nil
+	}
+	if !exists {
+		// nothing we need to do. It should always be possible to re-create this type of controller
+		c.expectations.DeleteExpectations(key)
+		return nil
+	}
+	OVM := obj.(*virtv1.OfflineVirtualMachine)
+
+	logger := log.Log.Object(OVM)
+
+	logger.Infof("Working on OVM: %s", OVM.Name)
+
+	//TODO default rs if necessary, the aggregated apiserver will do that in the future
+	if OVM.Spec.Template == nil || OVM.Spec.Selector == nil || len(OVM.Spec.Template.ObjectMeta.Labels) == 0 {
+		logger.Error("Invalid controller spec, will not re-enqueue.")
+		return nil
+	}
+
+	selector, err := v1.LabelSelectorAsSelector(OVM.Spec.Selector)
+	if err != nil {
+		logger.Reason(err).Error("Invalid selector on replicaset, will not re-enqueue.")
+		return nil
+	}
+
+	if !selector.Matches(labels.Set(OVM.Spec.Template.ObjectMeta.Labels)) {
+		logger.Reason(err).Error("Selector does not match template labels, will not re-enqueue.")
+		return nil
+	}
+
+	needsSync := c.expectations.SatisfiedExpectations(key)
+
+	// get all potentially interesting VMs from the cache
+	vms, err := c.listVMsFromNamespace(OVM.ObjectMeta.Namespace)
+
+	if err != nil {
+		logger.Reason(err).Error("Failed to fetch vms for namespace from cache.")
+		return err
+	}
+
+	vms = c.filterActiveVMs(vms)
+
+	// If any adoptions are attempted, we should first recheck for deletion with
+	// an uncached quorum read sometime after listing VirtualMachines (see kubernetes/kubernetes#42639).
+	canAdoptFunc := controller.RecheckDeletionTimestamp(func() (v1.Object, error) {
+		fresh, err := c.clientset.OfflineVirtualMachine(OVM.ObjectMeta.Namespace).Get(OVM.ObjectMeta.Name, &v1.GetOptions{})
+		if err != nil {
+			return nil, err
+		}
+		if fresh.ObjectMeta.UID != OVM.ObjectMeta.UID {
+			return nil, fmt.Errorf("original OfflineVirtualMachine %v/%v is gone: got uid %v, wanted %v", OVM.Namespace, OVM.Name, fresh.UID, OVM.UID)
+		}
+		return fresh, nil
+	})
+	cm := controller.NewVirtualMachineControllerRefManager(controller.RealVirtualMachineControl{Clientset: c.clientset}, OVM, selector, virtv1.OfflineVirtualMachineGroupVersionKind, canAdoptFunc)
+	vms, err = cm.ClaimVirtualMachines(vms)
+	if err != nil {
+		return err
+	}
+	var vm *virtv1.VirtualMachine
+	// the OfflineVirtualMachine can manage ONLY one VirtualMachine
+	// if for some unknown reason this controller is managing more
+	// than one handle the error
+	if len(vms) > 1 {
```

Comment:
We can adopt based on the VM name (like defined in the spec). Then this condition is impossible.